src/term.rs

```diff
@@ -417,6 +418,73 @@ impl ArrayAttrs {
     }
 }
 
+pub mod record {
+    use super::{RecordAttrs, RichTerm};
+    use crate::identifier::Ident;
+    use std::collections::HashMap;
+
+    /// The base structure of a Nickel record.
+    ///
+    /// Used to group together fields common to both the `Record` and
+    /// `RecRecord` terms.
+    #[derive(Clone, Debug, PartialEq)]
+    pub struct Record {
+        /// Fields whose names are known statically.
+        pub fields: HashMap<Ident, RichTerm>,
+        /// Attributes which may be applied to a record.
+        pub attrs: RecordAttrs,
+    }
+
+    impl Record {
+        pub fn new(fields: HashMap<Ident, RichTerm>, attrs: RecordAttrs) -> Self {
+            Record { fields, attrs }
+        }
+
+        /// A record with no fields and the default set of attributes.
+        pub fn empty() -> Self {
+            let fields = HashMap::new();
+            let attrs = Default::default();
+            Record { fields, attrs }
+        }
+
+        /// A record with the provided fields & the default set of attributes.
+        pub fn with_fields(fields: HashMap<Ident, RichTerm>) -> Self {
+            let attrs = Default::default();
+            Record { fields, attrs }
+        }
+
+        /// Returns the record resulting from applying the provided function
+        /// to each field.
+        ///
+        /// Note that `f` is taken as `mut` in order to allow it to mutate
+        /// external state while iterating.
+        pub fn update_fields<F>(self, mut f: F) -> Record
+        where
+            F: FnMut(Ident, RichTerm) -> RichTerm,
+        {
+            self.filter_update_fields(|id, t| Some(f(id, t)))
+        }
+
+        /// Returns the record resulting from applying the provided function
+        /// to each field, and removing any field for which the function returns
+        /// None.
+        ///
+        /// Note that `f` is taken as `mut` in order to allow it to mutate
+        /// external state while iterating.
+        pub fn filter_update_fields<F>(self, mut f: F) -> Record
```

Comment:
Hmm, you're right, although we already mix static and runtime constructs (for example ops have an `eval_mode()` function that returns `eval::EvalMode`), it's not a great thing to do in general. That being said, we can still implement this function as a standalone helper somewhere inside the evaluation module (and get rid of `filter_update_fields`, if it's not used anymore) ?

Another solution is to use the trait `RecordBaseExt` pattern (use a boilerplate trait to add a function to `RecordBase` from within `eval`). But I think it's not really required here, because we don't need to access private data.

Reply:
I guess the only awkwardness this creates is that now both `map_fields` and the new standalone function have broadly the same implementation, but can't be deduplicated (i.e. `map_fields` can no longer call out to `filter_map_fields` for its implementation). But that's still a lot less duplication than before, so I guess I'll go with that. :grin: